src/SArray.jl

```diff
@@ -19,56 +19,82 @@ immutable SArray{Size, T, N, L} <: StaticArray{T, N}
         new{Size,T,N,L}(x)
     end
 
     function (::Type{SArray{Size,T,N,L}}){Size,T,N,L}(x::NTuple{L,Any})
-        check_sarray_parameters(Val{Size}, T, Val{N}, Val{L})
+        check_sarray_parameters(Size, T, Val{N}, Val{L})
         new{Size,T,N,L}(convert_ntuple(T, x))
     end
 end
 
-@generated function check_sarray_parameters{Size,T,N,L}(::Type{Val{Size}}, ::Type{T}, ::Type{Val{N}}, ::Type{Val{L}})
-    if !(isa(Size, Tuple{Vararg{Int}}))
-        error("SArray parameter Size must be a tuple of Ints (e.g. `SArray{(3,3)}`)")
+@pure tuple_length(T) = length(T.parameters)
+@pure tuple_prod(T) = *(T.parameters...)
+@pure tuple_minimum(T) = minimum(tuple(T.parameters...))
+
+# Something doesn't match up type wise
+function check_sarray_parameters(Size, T, N, L)
+    !isa(Size, DataType) || (Size.name !== Tuple.name) && throw(ArgumentError("SArray parameter Size must be a Tuple type, got $Size"))
+    !isa(T, Type) && throw(ArgumentError("SArray parameter T must be a type, got $T"))
+    !isa(N.parameters[1], Int) && throw(ArgumenError("SArray parameter N must be an integer, got $(N.parameters[1])"))
+    !isa(L.parameters[1], Int) && throw(ArgumentError("SArray parameter L must be an integer, got $(L.parameters[1])"))
+    # shouldn't reach here. Anything else should have made it to the function below
+    error("Internal error. Please file a bug")
+end
+
+@generated function check_sarray_parameters{Size,T,N,L}(::Type{Size}, ::Type{T}, ::Type{Val{N}}, ::Type{Val{L}})
+    if !all(x->isa(x, Int), Size.parameters)
+        throw(ArgumentError("SArray parameter Size must be a tuple of Ints (e.g. `SArray{Tuple{3,3}}` or `SMatrix{3,3}`)."))
     end
 
-    if L != prod(Size) || L < 0 || minimum(Size) < 0 || length(Size) != N
-        error("Size mismatch in SArray parameters. Got size $Size, dimension $N and length $L.")
+    if L != tuple_prod(Size) || L < 0 || tuple_minimum(Size) < 0 || tuple_length(Size) != N
+        throw(ArgumentError("Size mismatch in SArray parameters. Got size $Size, dimension $N and length $L."))
     end
 
     return nothing
 end
 
-@generated function (::Type{SArray{Size,T,N}}){Size,T,N}(x::Tuple)
+@generated function (::Type{SArray{Size,T,N}}){Size <: Tuple,T,N}(x::Tuple)
     return quote
         $(Expr(:meta, :inline))
-        SArray{Size,T,N,$(prod(Size))}(x)
+        SArray{Size,T,N,$(tuple_prod(Size))}(x)
     end
 end
 
-@generated function (::Type{SArray{Size,T}}){Size,T}(x::Tuple)
+@generated function (::Type{SArray{Size,T}}){Size <: Tuple,T}(x::Tuple)
     return quote
         $(Expr(:meta, :inline))
-        SArray{Size,T,$(length(Size)),$(prod(Size))}(x)
+        SArray{Size,T,$(tuple_length(Size)),$(tuple_prod(Size))}(x)
     end
 end
 
-@generated function (::Type{SArray{Size}}){Size, T <: Tuple}(x::T)
+function SArray{Size,T,N}(x::Tuple) where {Size, T, N}
+    Base.depwarn("SArray{(dim1,dim2,...),...} is deprecated, use SArray{Tuple{dim1, dim2, ...}, ...}", :SArray)
```

Comment:
Maybe. Arguably, this should just be an error immediately. The only concern here would be something that the type system doesn't let you put in Tuples. Happy to go with whatever you prefer.